File: scripts/import_ipam.rb
#!/usr/bin/env ruby
# frozen_string_literal: true
require 'optparse'
require 'tempfile'
require_relative 'policy_data'
directory = 'puppet/data'
OptionParser.new { |p| p.on('--data PATH') { |v| directory = v } }.parse!
abort 'Usage: import_ipam.rb [--data DIR] vendor-export.json' unless ARGV.length == 1
snapshot = JSON.parse(File.read(ARGV.first))
FirewallPoc::PolicyData.validate(FirewallPoc::PolicyData.load(directory, snapshot: snapshot))
output = File.join(directory, 'external')
FileUtils.mkdir_p(output)
Tempfile.create(['ipam', '.json'], output) do |file|
  file.write(JSON.pretty_generate('profile::policy::ipam' => snapshot) + "\n")
  file.flush
  File.rename(file.path, File.join(output, 'ipam.json'))
end
puts "Imported IPAM revision #{snapshot.fetch('revision')} into #{output}/ipam.json"

File: scripts/import_ipam.py
#!/usr/bin/env python3
"""Validate a vendor-neutral IPAM export and atomically publish it. No API credentials on VMs."""
import argparse
import json
from pathlib import Path
from compile_policy import compile_policy
p = argparse.ArgumentParser()
p.add_argument('export', type=Path)
p.add_argument('--config', type=Path, default=Path('config'))
a = p.parse_args()
snapshot = json.loads(a.export.read_text())
compile_policy(json.loads((a.config / 'inventory.json').read_text()),
               json.loads((a.config / 'policy.json').read_text()), snapshot)
path = a.config / 'ipam.json'
temp = path.with_suffix('.tmp')
temp.write_text(json.dumps(snapshot, indent=2) + '\n')
temp.replace(path)
print(f"Validated IPAM revision {snapshot['revision']}: {path}")
